bundles/org.openhab.binding.sensorpush/src/main/java/org/openhab/binding/sensorpush/internal/handler/CloudBridgeHandler.java

```diff
@@ -0,0 +1,438 @@
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Access request returned status " + Integer.valueOf(status));
+                return;
+            }
+
+            tokenResponse = gson.fromJson(accessContentResponse.getContentAsString(), AccessTokenResponse.class);
+            if (tokenResponse == null) {
+                throw new InvalidResponseException("Access token object is null");
+            }
+            String accessToken = tokenResponse.accessToken;
+            logger.trace("Received access token: {}", accessToken);
+
+            JwtInfo jwtInfo;
+            try {
+                jwtInfo = new JwtInfo(accessToken);
+            } catch (IllegalArgumentException e) {
+                throw new InvalidResponseException("Invalid JWT token");
+            }
+            this.accessToken = accessToken;
+            accessTokenInfo = jwtInfo;
+        } catch (
+
+        TimeoutException e) {
+            logger.debug("Timeout sending auth request");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Auth request timeout");
+            return;
+        } catch (InterruptedException e) {
+            logger.debug("Interrupted sending authorization request");
+            return;
+        } catch (ExecutionException e) {
+            logger.debug("Execution exception sending authorization request: {}", e.getMessage());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Auth request execution exception");
+            return;
+        } catch (JsonSyntaxException | InvalidResponseException e) {
+            logger.debug("Error parsing auth response: {}", e.getMessage());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Error parsing auth response");
+            return;
+        }
+    }
+
+    /**
+     * Returns true if the JWT token represented by jwtInfo is expired or will expire in the next 60 seconds.
+     */
+    private boolean tokenExpired(@Nullable JwtInfo jwtInfo) {
+        if (jwtInfo == null) {
+            return true;
+        } else {
+            return (jwtInfo.expires.minusSeconds(60).isBefore(Instant.now()));
+        }
+    }
+
+    /**
+     * Polling routine. Polls for samples every run, but sensors only every other run. This reduces the polling load on
+     * the cloud service.
+     */
+    private void doPolling() {
+        if (accessToken == null || accessTokenInfo == null || tokenExpired(accessTokenInfo)) {
+            authorize();
+        }
+
+        if (pollSensorsRun) {
+            pollSensors();
+            pollSensorsRun = false;
+        } else {
+            pollSensorsRun = true;
+        }
+
+        pollSamples();
+    }
+
+    /**
+     * Poll for sensor info. Also supplies sensor list to discovery.
+     */
+    public void pollSensors() {
+        logger.trace("Polling sensors");
```

Comment:
Can this message be replaced by using the debugger? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging
```
void myFun() {
    logger.trace("Enter myfun"); // DONT, DONT, really DONT do that
    doSomething();
    logger.trace("Leave myfun"); // DONT, DONT, really DONT do that
}
```

Please check all.

Reply:
I'd rather leave these trace level messages in critical places. Tracing with the debugger is fine when I'm running the binding myself, but when users report problems I know from past experience that being able to tell them to set the logging level to TRACE and post or email the log is worth a few extra lines of code.

Reply:
OK, but please keep them to a minimum.